server/build_event_protocol/target_tracker/target_tracker.go

```diff
@@ -163,16 +163,16 @@ func (t *TargetTracker) testTargetsInAtleastState(state targetState) bool {
 	return true
 }
 
-func (t *TargetTracker) writeTestTargets(ctx context.Context) error {
-	var permissions *perms.UserGroupPerm
+func (t *TargetTracker) permissionsFromContext(ctx context.Context) *perms.UserGroupPerm {
 	if auth := t.env.GetAuthenticator(); auth != nil {
 		if u, err := auth.AuthenticatedUser(ctx); err == nil && u.GetGroupID() != "" {
-			permissions = perms.GroupAuthPermissions(u.GetGroupID())
+			return perms.GroupAuthPermissions(u.GetGroupID())
 		}
 	}
-	if permissions == nil {
-		return status.FailedPreconditionError("Permissions were nil -- not writing target data.")
-	}
+	return nil
```

Comment:
I'm not a big fan of returning nil pointers from helper functions to indicate error as it could easily lead to a future nil dereference. 
I would change the function to return (*perms.UserGroupPerm, error)

Reply:
Done